mscorlib/system/datetime.cs

```diff
@@ -348,7 +348,12 @@ public DateTime(int year, int month, int day, Calendar calendar)
         // Returns the DateTime resulting from adding a fractional number of
         // time units to this DateTime.
         private DateTime Add(double value, int scale) {
-            long millis = (long)(value * scale + (value >= 0? 0.5: -0.5));
+            long millis;
+            try {
+                millis = checked((long)(value * scale + (value >= 0? 0.5: -0.5)));
+            } catch (OverflowException) {
+                throw new ArgumentOutOfRangeException("value", Environment.GetResourceString("ArgumentOutOfRange_AddValue"));
+            }
             if (millis <= -MaxMillis || millis >= MaxMillis) 
```

Comment:
@marek-safar yes, there are values that fit into long (and thus don't overflow) but are still larger/smaller than MaxMillis, so this is definitely needed.